src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -518,37 +518,43 @@ void HSHomeObject::on_blob_del_commit(int64_t lsn, sisl::blob const& header, sis
     RELEASE_ASSERT(index_table != nullptr, "Index table not intialized");
     RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
 
-    BlobInfo blob_info;
-    blob_info.shard_id = msg_header->shard_id;
-    blob_info.blob_id = *r_cast< blob_id_t const* >(key.cbytes());
+    const auto shard_id = msg_header->shard_id;
+    const auto blob_id = *r_cast< blob_id_t const* >(key.cbytes());
 
-    auto r = move_to_tombstone(index_table, blob_info);
-    if (!r) {
-        if (recovery_done_) {
-            BLOGE(tid, blob_info.shard_id, blob_info.blob_id, "Failed to move blob to tombstone, error={}", r.error());
-            if (ctx) ctx->promise_.setValue(folly::makeUnexpected(r.error()));
-        } else {
-            if (ctx) ctx->promise_.setValue(BlobManager::Result< BlobInfo >(blob_info));
-        }
+    BlobRouteKey index_key{BlobRoute{shard_id, blob_id}};
+    BlobRouteValue index_value{tombstone_pbas};
+    BlobRouteValue existing_value;
 
-        // if we return directly, for the perspective of statemachin, this lsn has been committed successfully. so there
-        // will be no more deletion for this blob, and as a result , blob leak happens
-        RELEASE_ASSERT(false, "fail to commit delete blob log for pg={}, shard={}, blob={}", msg_header->pg_id,
-                       msg_header->shard_id, blob_info.blob_id);
+    homestore::BtreeSinglePutRequest update_req{&index_key, &index_value, homestore::btree_put_type::UPDATE,
+                                                &existing_value};
+
+    auto status = index_table->put(update_req);
+    if (status != homestore::btree_status_t::success) {
+        // if we return directly, for the perspective of statemachine, this lsn has been committed successfully. so
+        // there will be no more deletion for this blob, and as a result , blob leak happens
+        RELEASE_ASSERT(false, "fail to commit delete_blob log for pg={}, shard={}, blob={}, lsn={}", msg_header->pg_id,
+                       msg_header->shard_id, blob_id, lsn);
     }
 
-    LOGD("shard_id={}, blob_id={} has been moved to tombstone, lsn={}", blob_info.shard_id, blob_info.blob_id, lsn);
+    LOGD("shard_id={}, blob_id={} has been moved to tombstone, lsn={}", shard_id, blob_id, lsn);
 
-    auto& multiBlks = r.value();
-    if (multiBlks != tombstone_pbas) {
-        repl_dev->async_free_blks(lsn, multiBlks);
-        const_cast< HS_PG* >(hs_pg)->durable_entities_update([](auto& de) {
-            de.active_blob_count.fetch_sub(1, std::memory_order_relaxed);
-            de.tombstone_blob_count.fetch_add(1, std::memory_order_relaxed);
+    auto existing_pbas = existing_value.pbas();
```

Comment:
This is an behavior change that the response of deleting a non-exists blob changed from UNKNOWN_BLOB to SUCCESS.

Also , if the blob doesnt exists , the existing_pba will be default value of all zero.  is that right to free in this case?

I was surprise there is no UT covered this but lets add one.

Reply:
I have updated the code to cover the case that blob does not exist, will try to add a UT later